### ISSUE TITLE
Incompatible with sexplib/ppx_sexp_conv v0.11.0

### ISSUE THREAD

**Issue by copy** [mirleft/ocaml-nocrypto#143]
After upgrading sexplib and ppx_sexp_conv to v0.11.0 the following error appears when compiling nocrypto:

```
Error: The implementation (obtained by packing)
       does not match the interface src/nocrypto.mli:
       ...
       In module Hash:
       Values do not match:
         val hash_of_sexp : Ppx_sexp_conv_lib.Sexp.t -> hash
       is not included in
         val hash_of_sexp : Sexplib.Sexp.t -> hash
       File "src/nocrypto.mli", line 217, characters 2-43:
         Expected declaration
       File "src/hash.ml", line 114, characters 0-86: Actual declaration
```

**Comment by gasche** [comment]
I just noticed the same error and reported a bug upstream at `ppx_sexp_conv`: https://github.com/janestreet/ppx_sexp_conv/issues/20 .

**Comment by hannesm** [comment]
FWIW: diml has an explanation for this in https://github.com/ocaml/opam-repository/pull/11628#issuecomment-375697444

I've "fixed" this issue in https://github.com/ocaml/opam-repository/pull/11636 by adding an upper bound of sexplib (`<"v0.11.0"`) to nocrypto releases.  of course a proper fix and a release of nocrypto would be welcome.

**Comment by gasche** [comment]
Of course my curiosity was piqued as it revealed to be a build system issue. I have two bugfix branches, [master-ocamlbuild-pack](https://github.com/mirleft/ocaml-nocrypto/compare/master...gasche:master-ocamlbuild-pack) and [v0.5.4-ocamlbuild-pack](https://github.com/mirleft/ocaml-nocrypto/compare/v0.5.4...gasche:v0.5.4-ocamlbuild-pack).

@hannesm: if you know how to get one of these in an OPAM-available release of nocrypto, that could be nice (the v0.5.4 branch would be suitable for a `patch/`-level fix, but I understand that we are not excited with opam-level patching anymore). I will get the ocamlbuild-side issue fixed, but an ocamlbuild update will not fix the nocrypto build, it still needs a change to the `_tags` file.

**Comment by gasche** [comment]
@hcarty: yes, that seems acceptable to me given nocrypto's release management style. Would you be willing to submit a PR to opam-repository?

**Comment by hannesm** [comment]
(no, I'm neither author nor maintainer of nocrypto) tbh, I'm really against depending on the ppx machinery at runtime, since this will increase the binary size of all my unikernels tremendously.  in the past I've been removing runtime dependencies of various ppx in the MirageOS ecosystem.  I'm not sure how to move forward, to me it looks like getting rid of ppx_sexp_conv entirely is the only viable solution.

**Comment by gasche** [comment]
If you already depend on Base, there is no additional size cost. If you don't, then yes, you may want to use a serializer/deserializer that comes with smaller runtime dependencies (it may be possible to have something that is drop-in compatible with `with sexp` but with a smaller runtime).

**Comment by hannesm** [comment]
Dear @gasche, there is no dependency on `base` in the MirageOS ecosystem (sexplib 0.9.0 required base, which increased the binary size by 5MB of every unikernel, which was fixed by the sexplib maintainers).

**Comment by gasche** [comment]
Well I guess that you should complain to the sexplib maintainers again, as I suppose that they have some control over how ppx_sexplib sets its runtime dependencies -- it should be possible to only depend on `sexplib`.

**Comment by gasche** [comment]
cc @trefis :-)

**Comment by ansiwen** [comment]
I want to mention that my project uses the serializers, so please don’t remove them without replacement. Isn’t it possible to put them in a separate package, so they are only added if used?

Anyway I don’t understand, why the linker can’t remove unused code?

**Comment by gasche** [comment]
The linker can remove unused compilation units (toplevel modules), but doesn't go at a finer-grained level, and it is easy to have (unused) definitions in your (used) module pull out extra undesirable dependency. A global tree-shaker / dead code remover link-time pass could be implemented in theory (and is an interesting subject, particularly for unikernels), but it is difficult (in theory and in practice) due the presence of side-effect-performing expressions at the toplevel of modules (can those initializing an otherwise-unused datastructure be also removed, or is their effect observable by other parts of the codebase?).

**Comment by cfcs** [comment]
Has anyone pinged @pqwy about this? I don't know how to contact him now that he also stopped responding to emails, but this is an annoying issue to have to work around.

**Comment by j0sh** [comment]
As far as I can tell, we can't use `nocrypto` along with a `ppxlib` driven rewriter. The alternative is using the deprecated `ppx_type_conv` for rewriters. Would really appreciate a long-term fix here.

**Comment by copy** [comment]
This has been fixed using opam patches in https://github.com/ocaml/opam-repository/pull/12062 and released as nocrypto 0.5.4-1.

**Comment by cfcs** [comment]
should be fixed in latest master: https://github.com/mirleft/ocaml-nocrypto/commit/a58c6534c5cc5da1598f8c979ea397ea2c6b92e1
